### ISSUE TITLE
Epipolar and homography error metrics add `eps` to the value: Sampson 90% low at ‖F‖ = 1e-3, exact matches score 1e-4

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4881]
The two-view error metrics add `eps` inside their arithmetic instead of guarding the one singular case. `sampson_epipolar_distance` and `symmetrical_epipolar_distance` add `eps` (default 1e-8) to the squared epiline norms in the denominator and, with `squared=False`, inside the square root; `left_to_right_epipolar_distance` / `right_to_left_epipolar_distance` go through `point_line_distance`, which adds `1e-9` to the epiline norm; `oneway_transfer_error` / `symmetric_transfer_error` add `eps` to the projective denominator and inside the square root. The docstrings describe `eps` as "Small constant for safe sqrt." Three consequences:

- **(a) The epipolar distances depend on the scale of `F`**, which is defined only up to scale. With 1-px noise, Sampson is 90% below the Hartley–Zisserman value (eq. 11.9) at `‖F‖ = 1e-3`, 1.1% below it at the scale `fundamental_from_essential` returns for this pair (`‖F‖ = 0.029`), and 9.3e-6 at `‖F‖ = 1`; `symmetrical` is about twice as far off; the one-way distances much less, since `point_line_distance` adds a smaller `1e-9` to the unsquared norm. An exact fixture cannot show this: the numerator is 0 there, whatever the denominator.
- **(b) `squared=False` scores an exact correspondence `sqrt(eps) = 1e-4`**, not 0, in both epipolar metrics and both homography transfer errors.
- **(c) CUDA and CPU disagree.** On a CUDA tensor with `N < use_matmul_at_less_than_points` (default 10000), `sampson_epipolar_distance` dispatches to `_sampson_epipolar_distance_matmul_impl_`, whose denominator (`_metrics.py:130`) has no `eps`; CPU, MPS, and CUDA with `N ≥ 10000` use `_sampson_epipolar_distance_manual_impl_`, which adds it (`_metrics.py:84`). The two differ by the scale-dependent amount of (a), 1.1% at the `fundamental_from_essential` scale, and on a degenerate `F` the manual path returns 1e8 where the matmul path returns `inf`. `symmetrical_epipolar_distance` also has two paths, but both add `eps` in the same places. **Not executed on CUDA** (none available): (c) is read from the dispatch code and shown by calling the two private implementations on CPU.

Every `+ eps` site:

| file:line | function | where |
|---|---|---|
| `kornia/geometry/epipolar/_metrics.py:84` | `_sampson_epipolar_distance_manual_impl_` | denominator |
| `kornia/geometry/epipolar/_metrics.py:88` | `_sampson_epipolar_distance_manual_impl_` | `sqrt` (`squared=False`) |
| `kornia/geometry/epipolar/_metrics.py:134` | `_sampson_epipolar_distance_matmul_impl_` | `sqrt` (`squared=False`); its denominator, line 130, has none |
| `kornia/geometry/epipolar/_metrics.py:212`, `:213` | `_symmetrical_epipolar_distance_manual_impl_` | both inverse denominators |
| `kornia/geometry/epipolar/_metrics.py:219` | `_symmetrical_epipolar_distance_manual_impl_` | `sqrt` |
| `kornia/geometry/epipolar/_metrics.py:236`, `:237` | `_symmetrical_epipolar_distance_matmul_impl_` | both inverse denominators |
| `kornia/geometry/epipolar/_metrics.py:242` | `_symmetrical_epipolar_distance_matmul_impl_` | `sqrt` |
| `kornia/geometry/linalg.py:275` | `point_line_distance` (`eps=1e-9`; used by `left_to_right_` / `right_to_left_epipolar_distance`, `_metrics.py:301`, `:334`) | denominator (the norm, not squared) |
| `kornia/geometry/homography.py:87`, `:88` | `oneway_transfer_error` | projective denominator `w' + eps` |
| `kornia/geometry/homography.py:94` | `oneway_transfer_error` | `sqrt` |
| `kornia/geometry/homography.py:141` | `symmetric_transfer_error` | `sqrt` (its two one-way terms, `:135`, `:136`, carry the `:87`/`:88` sites) |

## Repro

```python
import torch
from kornia.geometry.conversions import axis_angle_to_rotation_matrix
from kornia.geometry.epipolar import (essential_from_Rt, fundamental_from_essential, left_to_right_epipolar_distance,
                                      right_to_left_epipolar_distance, sampson_epipolar_distance,
                                      symmetrical_epipolar_distance)
from kornia.geometry.epipolar._metrics import (_sampson_epipolar_distance_manual_impl_,
                                               _sampson_epipolar_distance_matmul_impl_)
from kornia.geometry.homography import oneway_transfer_error, symmetric_transfer_error

d = torch.float64
K1 = torch.tensor([[[800.0, 0, 320], [0, 760, 200], [0, 0, 1]]], dtype=d)
K2 = torch.tensor([[[700.0, 0, 300], [0, 740, 240], [0, 0, 1]]], dtype=d)
R = axis_angle_to_rotation_matrix(torch.tensor([[0.1, -0.2, 0.05]], dtype=d))
t = torch.tensor([[[0.5], [0.1], [0.02]]], dtype=d)
torch.manual_seed(0)
X = torch.rand(12, 3, dtype=d) * 2 - 1
X[:, 2] += 5
x1 = (K1[0] @ X.T).T
x2 = (K2[0] @ (R[0] @ X.T + t[0])).T
x1, x2 = (x1[:, :2] / x1[:, 2:])[None], (x2[:, :2] / x2[:, 2:])[None]
x2n = x2 + torch.randn(1, 12, 2, dtype=d)                   # 1 px noise in image 2
Fe = fundamental_from_essential(essential_from_Rt(torch.eye(3, dtype=d)[None], torch.zeros(1, 3, 1, dtype=d), R, t), K1, K2)

def ref(F, a, b):
    """scale-free H&Z formulas: Sampson, symmetric (sum of squares), and the two one-way distances"""
    ah, bh = torch.cat([a, torch.ones_like(a[..., :1])], -1), torch.cat([b, torch.ones_like(b[..., :1])], -1)
    l2, l1 = ah @ F.transpose(-2, -1), bh @ F                # epilines in image 2 and in image 1
    r = (bh * l2).sum(-1)
    n2, n1 = l2[..., :2].pow(2).sum(-1), l1[..., :2].pow(2).sum(-1)
    return r**2 / (n2 + n1), r**2 * (1 / n2 + 1 / n1), r.abs() / n2.sqrt(), r.abs() / n1.sqrt()

print("(a) median relative error vs the scale-free formula, 1-px noise, float64:")
print(f"{'||F||':>12s} {'sampson':>10s} {'symmetric':>10s} {'left_to_right':>14s} {'right_to_left':>14s}")
rows = [(s, s * Fe / Fe.norm()) for s in (1e-3, 1e-1, 1.0, 1e2, 1e3)] + [("f_from_E", Fe)]
for s, F in rows:
    got = (sampson_epipolar_distance(x1, x2n, F), symmetrical_epipolar_distance(x1, x2n, F),
           left_to_right_epipolar_distance(x1, x2n, F), right_to_left_epipolar_distance(x1, x2n, F))
    err = [((g - r).abs() / r).median().item() for g, r in zip(got, ref(F, x1, x2n))]
    lab = f"{s:>12g}" if not isinstance(s, str) else f"{F.norm().item():>7.3g} (*)"
    print(f"{lab} {err[0]:>10.2g} {err[1]:>10.2g} {err[2]:>14.2g} {err[3]:>14.2g}")
print("(*) F = fundamental_from_essential(essential_from_Rt(...), K1, K2), unnormalised")

print("(b) exact correspondences, squared=False:")
print("   sampson min", sampson_epipolar_distance(x1, x2, Fe, squared=False).min().item(),
      "| with eps=0:", sampson_epipolar_distance(x1, x2, Fe, squared=False, eps=0.0).max().item())
print("   symmetrical min", symmetrical_epipolar_distance(x1, x2, Fe, squared=False).min().item())
H = torch.tensor([[[1.1, 0.05, 20.0], [-0.03, 0.95, -10.0], [1e-4, -2e-4, 1.0]]], dtype=d)
y = torch.cat([x1, torch.ones_like(x1[..., :1])], -1) @ H.transpose(-2, -1)
y = y[..., :2] / y[..., 2:]                                  # exact homography matches
print("   oneway_transfer_error min", oneway_transfer_error(x1, y, H, squared=False).min().item(),
      "| with eps=0:", oneway_transfer_error(x1, y, H, squared=False, eps=0.0).max().item())
print("   symmetric_transfer_error min", symmetric_transfer_error(x1, y, H, squared=False).min().item())

print("(c) the two private Sampson implementations (the public function picks matmul on CUDA for N < 10000):")
for name, F in [("F / ||F||", Fe / Fe.norm()), ("f_from_E", Fe), ("1e-3 F / ||F||", 1e-3 * Fe / Fe.norm())]:
    man = _sampson_epipolar_distance_manual_impl_(x1, x2n, F)
    mat = _sampson_epipolar_distance_matmul_impl_(x1, x2n, F)
    print(f"   {name:15s} median |matmul - manual| / matmul = {((mat - man).abs() / mat).median().item():.2g}")
Fdeg = torch.zeros(1, 3, 3, dtype=d)
Fdeg[0, 2, 2] = 1.0                                          # (F x)_{1:2} = 0 for every point: the singular case eps guards
print("   degenerate F (only F[2,2] = 1): manual", _sampson_epipolar_distance_manual_impl_(x1, x2n, Fdeg)[0, 0].item(),
      "| matmul", _sampson_epipolar_distance_matmul_impl_(x1, x2n, Fdeg)[0, 0].item())
```

Output:

```text
(a) median relative error vs the scale-free formula, 1-px noise, float64:
       ||F||    sampson  symmetric  left_to_right  right_to_left
       0.001        0.9       0.95        4.3e-05        4.3e-05
         0.1    0.00093     0.0019        4.3e-07        4.3e-07
           1    9.3e-06    1.9e-05        4.3e-08        4.3e-08
         100    9.3e-10    1.9e-09        4.3e-10        4.3e-10
        1000    9.3e-12    1.9e-11        4.3e-11        4.3e-11
 0.0293 (*)      0.011      0.021        1.5e-06        1.5e-06
(*) F = fundamental_from_essential(essential_from_Rt(...), K1, K2), unnormalised
(b) exact correspondences, squared=False:
   sampson min 0.0001 | with eps=0: 4.3830699160301277e-14
   symmetrical min 0.0001
   oneway_transfer_error min 0.00010004688338524099 | with eps=0: 5.684341886080802e-14
   symmetric_transfer_error min 0.00010008267291590023
(c) the two private Sampson implementations (the public function picks matmul on CUDA for N < 10000):
   F / ||F||       median |matmul - manual| / matmul = 9.3e-06
   f_from_E        median |matmul - manual| / matmul = 0.011
   1e-3 F / ||F||  median |matmul - manual| / matmul = 0.9
   degenerate F (only F[2,2] = 1): manual 100000000.0 | matmul inf
```

kornia `main` 7ddf731bd, torch 2.14.0 (and 2.5.1: same, except rounding in the last digit of two values: `9.4e-12` in the `1000` row, and `symmetric_transfer_error min 0.00010008267292119787`), CPU float64, macOS arm64.

A fix cannot just set `eps = 0`: at a point exactly on its epiline the unsquared Sampson distance then has a `NaN` gradient, where the default `eps` returns a zero gradient.

```python
import torch
from kornia.geometry.epipolar import sampson_epipolar_distance
F = torch.tensor([[[0.0, -0.02, 0.3], [0.02, 0.0, -0.9], [-0.3, 0.9, 0.1]]], dtype=torch.float64)
p1 = torch.tensor([[[10.0, 20.0]]], dtype=torch.float64)
l = torch.cat([p1, torch.ones(1, 1, 1, dtype=torch.float64)], -1) @ F.transpose(-2, -1)   # epiline of p1 in image 2
x = 5.0
y = -(l[0, 0, 0] * x + l[0, 0, 2]) / l[0, 0, 1]
p2 = torch.tensor([[[x, y.item()]]], dtype=torch.float64, requires_grad=True)          # exactly on the line
for eps in (1e-8, 0.0):
    v = sampson_epipolar_distance(p1, p2, F, squared=False, eps=eps)
    (g,) = torch.autograd.grad(v.sum(), p2)
    print(f"eps={eps:g}: residual^2 {sampson_epipolar_distance(p1, p2, F, eps=0.0).item():.3g}, value {v.item():.3g}, grad {g.flatten().tolist()}")
```

Output:

```text
eps=1e-08: residual^2 0, value 0.0001, grad [0.0, 0.0]
eps=0: residual^2 0, value 0, grad [nan, nan]
```

kornia `main` 7ddf731bd, torch 2.14.0 (and 2.5.1: same), CPU float64, macOS arm64.

## Classification

Silent wrong result. Idiom 1 (`x + eps` perturbing a result instead of guarding a singularity); (c) is the same idiom with the `eps` placed differently on each device path.

## Disposition

MERGE IN WINDOW — behaviour change: take `eps` out of the value. Guard only the singular denominators (an epiline with `a = b = 0`, `w' = 0`) with `torch.where` and a safe substitute, and take the square root through a safe argument so that an exact match gives 0 with a finite gradient; the matmul and manual paths then agree. Users see: the epipolar distances no longer change with the scale of `F` (percent-level changes and more for `‖F‖ ≲ 1e-2`, about 1e-5 relative at `find_fundamental`'s `F[2,2] = 1` scale), `squared=False` returns 0 for an exact match in all four functions, and CUDA and CPU return the same Sampson values.

Found by the batch-8 conventions audit (two-view geometry); the `test_wart_*` pins in the docs PR flip when the fix lands.

Related: #4874 (same idiom in normalize_transformation).

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).
